src/overmap.cpp

```diff
@@ -4168,6 +4186,71 @@ std::vector<point> overmap::get_sectors() const
     return res;
 }
 
+bool overmap::place_special_attempt( std::vector<std::pair<const overmap_special *, int>> &candidates,
+                                     const point &sector, bool check_cities )
+{
+    const int x = sector.x;
+    const int y = sector.y;
+
+    const tripoint p( rng( x, x + OMSPEC_FREQ - 1 ), rng( y, y + OMSPEC_FREQ - 1 ), 0 );
+    const city &nearest_city = get_nearest_city( p );
+
+    for( auto iter = candidates.begin(); iter != candidates.end(); ++iter ) {
+        const auto &special = *iter->first;
+        // City check is the fastest => it goes first.
+        if( check_cities && !special.can_belong_to_city( p, nearest_city ) ) {
+            continue;
+        }
+        // See if we can actually place the special there.
+        const auto rotation = random_special_rotation( special, p );
+        if( rotation == om_direction::type::invalid ) {
+            continue;
+        }
+
+        place_special( special, p, rotation, nearest_city );
+
+        if( --iter->second == 0 ) {
+            if( candidates.empty() ) {
+                return true; // Job done. Bail out.
+            }
+            iter = candidates.erase( iter );
+        }
+
+        return true;
+    }
+
+    return false;
+}
+
+void overmap::place_specials_pass( std::vector<std::pair<const overmap_special *, int>> &to_place,
+                                   std::vector<point> &sectors, bool check_cities )
+{
+    // Walk over sectors in random order, to minimize "clumping"
+    std::random_shuffle( sectors.begin(), sectors.end() );
+    for( auto it = sectors.begin(); it != sectors.end(); ) {
+        const point &cur_sector = *it;
+        const size_t attempts = 10;
+        bool placed = false;
+        for( size_t i = 0; i < attempts; ++i ) {
+            if( place_special_attempt( to_place, cur_sector, check_cities ) ) {
+                placed = true;
+                it = sectors.erase( it );
```

Comment:
Why do you erase the current sector? Since the code iterates only forward, that doesn't seem to be needed (especially given the expensiveness of shifts in the vector).